Initial Hyperstack cloud implementation

vvidovic opened the pull request:
Initial implementation of the SkyPilot provisioning for the Hyperstack cloud.

<!-- Describe the changes in this PR -->



<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

vvidovic-croz committed:
Initial Hyperstack cloud implementation

Initial implementation of the SkyPilot provisioning for the
Hyperstack cloud.

cblmemo commented on sky/provision/hyperstack/hyperstack_utils.py:
```diff
@@ -0,0 +1,224 @@
+        default_environment = f'default-{region}'
+
+        ssh_key = self.get_or_add_ssh_key(default_environment, ssh_pub_key)
+        security_rules = [
+            dict(
+                direction='ingress',
+                protocol='tcp',
+                ethertype='IPv4',
+                remote_ip_prefix='0.0.0.0/0',
+                port_range_min=22,
+                port_range_max=22,
+            )
+        ]
+        if ports:
+            for p in ports:
+                security_rules.append(
+                    dict(
+                        direction='ingress',
+                        protocol='tcp',
+                        ethertype='IPv4',
+                        remote_ip_prefix='0.0.0.0/0',
+                        port_range_min=p,
+                        port_range_max=p,
+                    ))
```
can we have a function to generate one single rule and reuse it?

vvidovic replied:
Do you maybe know if there is a SkyPilot configuration approach that would enable us to set the `remote_ip_prefix` (the IP address range from which it is allowed to connect to the port)?
(sorry for the digression)

vvidovic replied:
done.

cblmemo replied:
Nope - currently we assume all ip can access the skypilot's vm. To confirm, this is essentially the outbound port, right?

cblmemo replied:
Filed an issue #5288 to keep track of it. Feel free to discuss any related topics under that issue :))

vvidovic replied:
I didn't mean the outbound port, but the outbound IP address.
It is just an additional check on the VM side that it is not available to the entire Internet but only to some specific IP or IP range of addresses.

For example, you could want to allow access from the:
- AWS (because some other component of your system is hosted in AWS)
- your company VPN (for example, you run some internal LLM model in the cloud, but want to allow only your employees to connect to it)

cblmemo replied:
Got it. Currently we dont have this limitation built-in. Just reposted this in the issue and I think we can skip this limit in this PR :))

cblmemo commented on sky/provision/hyperstack/hyperstack_utils.py:
```diff
@@ -0,0 +1,224 @@
+
+        return instance_ids
+
+    def list_ssh_keys(self, environment_name: str):
```
type annotation for the return value?

vvidovic replied:
Added `List[dict]`. Done.

cblmemo commented on sky/provision/hyperstack/instance.py:
```diff
@@ -0,0 +1,339 @@
+
+def _filter_instances(
+        cluster_name_on_cloud: str,
+        add_active: bool = True,
+        add_pending: bool = True,
+        add_stopped: bool = True,
```
This is a little bit confusing. Can we have a status filter like this?

https://github.com/skypilot-org/skypilot/blob/134b8feb0fcbe0a0d1763592a4186dd172e9f41a/sky/provision/runpod/instance.py#L19-L35

vvidovic replied:
Maybe I can add a "status groups" to filter?
(`PENDING`, `ACTIVE`, `STOPPED`)

cblmemo replied:
IIUC the "status group to filter" is exactly the `status_filters: Optional[List[str]]` argument in the above function? Could you elaborate on the differences?

vvidovic replied:
There are 9 recognised Hyperstack statuses (from the Hypercloud responses or web UI):
- 1 of those is `ACTIVE`
- 1 of those is `STOPPED`
- all other statuses are `PENDING` (`BUILD`, `CREATING`, `STARTING`, `REBOOTING`, `HARD_REBOOT` & `DELETING`)

That means that each time we would like to filter by Hyperstack status group "PENDING" we would need to add all 6 of those to the list and when we want to combine for example "ACTIVE" and "PENDING" it would be harder to understand that we want "all pending statuses" + "active status".

cblmemo replied:
Got it. Then adding a status group to filter sounds great to me!

cblmemo commented on sky/provision/hyperstack/instance.py:
```diff
@@ -0,0 +1,339 @@
+                                          False)
+    instances: Dict[str, List[common.InstanceInfo]] = {}
+
+    # subprocess_utils.run_in_parallel(get_internal_ip,
+    #                                  list(running_instances.values()))
+    head_instance_id = None
```
```suggestion
    head_instance_id = None
```

accidental?

vvidovic replied:
done.

cblmemo reviewed (commented):
Thanks for adding this @vvidovic ! This is exciting. Left some discussions.

Also, could you help run smoke test on hyperstack to make sure it works properly?

vvidovic commented on sky/backends/backend_utils.py:
```diff
@@ -897,7 +898,7 @@ def _add_auth_to_cluster_config(cloud: clouds.Cloud, cluster_config_file: str):
         config = auth.setup_fluidstack_authentication(config)
     else:
-        assert False, cloud
+        assert False, f'{cloud} - adding authentication to cluster config not implemented.'
```
In the process of adding the Hyperstack provider, there were many errors with non-descriptive messages. I was thinking that this kind of more descriptive error message could help future implementations to easier find where from the error was thrown.

Of course, it is quite easy to remove that if this argument doesn't make sense.

cblmemo replied:
Oh, good catch!

I thought that since here we are asserting false, this code path should not be entered and thus adding extra information is not of big use. But if you actually encountered this code path, a valid choice would be raising some error instead - that would be more informative. Could you help changing this to a `ValueError` or sth? thanks!

vvidovic commented on sky/clouds/hyperstack.py:
```diff
@@ -0,0 +1,313 @@
+        Hyperstack's storage service."""
+        return cls._check_credentials()
+
+    @classmethod
+    def _check_credentials(cls) -> Tuple[bool, Optional[str]]:
+        try:
+            assert os.path.exists(
+                os.path.expanduser(hyperstack_utils.HYPERSTACK_API_KEY_PATH))
+        except AssertionError:
+            return False, ('Failed to access Hyperstack Cloud'
+                           ' with credentials. '
+                           'To configure credentials, go to:\n    '
+                           '  https://console.hyperstack.cloud/api-keys \n    '
+                           'to obtain an API key, '
+                           'then add save the contents '
+                           'to ~/.hyperstack/api_key \n')
+        except requests.exceptions.ConnectionError:
+            return False, ('Failed to verify Hyperstack Cloud credentials. '
+                           'Check your network connection '
+                           'and try again.')
+        return True, None
```
It seems I improperly copied some other cloud logic, let me fix that.

I believe the Fluidstack check is the one this was based on:
- https://github.com/skypilot-org/skypilot/blob/34973e2088e86d011a925e5dabd7b150a297eee6/sky/clouds/fluidstack.py#L261

vvidovic replied:
done.

vvidovic commented on sky/clouds/hyperstack.py:
```diff
@@ -0,0 +1,313 @@
+    def regions(cls) -> List['clouds.Region']:
+        return service_catalog.regions(clouds='hyperstack')
+
+    @classmethod
+    def query_status(cls, name: str, tag_filters: Dict[str, str],
+                     region: Optional[str], zone: Optional[str],
+                     **kwargs) -> List[status_lib.ClusterStatus]:
+        status_map = {
+            'booting': status_lib.ClusterStatus.INIT,
+            'active': status_lib.ClusterStatus.UP,
+            'unhealthy': status_lib.ClusterStatus.INIT,
+            'terminating': None,
+            'terminated': None,
+        }
+        # TODO(ewzeng): filter by hash_filter_string to be safe
+        status_list = []
+        vms = hyperstack_utils.HyperstackClient().list_instances()
+        possible_names = [f'{name}-head', f'{name}-worker']
+        for node in vms:
+            if node.get('name') in possible_names:
+                node_status = status_map[node['status']]
+                if node_status is not None:
+                    status_list.append(node_status)
+        return status_list
```
True, this method was not used and statuses were not correct for the Hyperstack.

vvidovic replied:
done.

vvidovic commented on sky/clouds/service_catalog/data_fetchers/fetch_hyperstack.py:
```diff
@@ -0,0 +1,147 @@
+                                    [--api-key-path API_KEY_PATH]
+
+If neither --api-key nor --api-key-path are provided, this script will parse
+`~/.hyperstack/credentials.yaml` to look for IBM API key (`iam_api_key`).
```
It was an outdated comment copied from the IBM cloud fetcher.

How do you mean to add a precedence? The first one found is used:
- `api_key` arg
- `api_key_path` arg
- default API key path

I think that should be fine.

vvidovic replied:
done.

cblmemo replied:
sry, i mean adding the precedence in the comments, which is basically this list's order.

vvidovic commented on sky/clouds/service_catalog/data_fetchers/fetch_hyperstack.py:
```diff
@@ -0,0 +1,147 @@
+                      encoding='utf-8') as f:
+                api_key_res = f.read().strip()
+        else:
+            # Read from ~/.lambda_cloud/lambda_keys
```
done.

vvidovic commented on sky/provision/hyperstack/hyperstack_utils.py:
```diff
@@ -0,0 +1,224 @@
+            flavor for group in flavor_groups for flavor in group['flavors']
+        ]
+
+        return flavors
```
Added `List[dict]`. Done.

vvidovic commented:
@cblmemo - thanks for the review. There are 2 issues left:
- more descriptive error message (remove or not)
- simplify filtering instances method

I made a few other changes:
- added type annotations where I noticed that these are missing
- added an additional check and warning for a missing price for a VM flavor (Hyperstack added the `H200-141G-SXM5` GPU but didn't add it to the price list - https://github.com/vvidovic/skypilot/blob/a037fe9637d2f555e8c8cf01096eeaf1ee27e225/sky/clouds/service_catalog/data_fetchers/fetch_hyperstack.py#L109).

I implemented changes using a force push to my fork. This has a downside in that there is no history of changes I performed based on your review, but it makes the future merge simpler.
If you don't like this approach, I will continue using normal commits and maybe rebase them at the end (for a simpler merge).

vvidovic commented:
> Thanks for adding this @vvidovic ! This is exciting. Left some discussions.
> 
> Also, could you help run smoke test on hyperstack to make sure it works properly?

I have run the following tests:
```
# Check credentials:
python sky/cli.py check

python sky/cli.py launch -c test-single-instance --cloud hyperstack --gpus RTX-A6000 echo hi
python sky/cli.py stop test-single-instance
python sky/cli.py start test-single-instance

python sky/cli.py launch -c test-single-instance echo hi
python sky/cli.py down test-single-instance

# Terminate a stopped instance:
python sky/cli.py launch -c test-single-instance --cloud hyperstack --gpus RTX-A6000 echo hi
python sky/cli.py stop test-single-instance; sky down test-single-instance

python sky/cli.py launch --cloud hyperstack --gpus RTX-A6000 -c test-autostop -i 1 echo hi
python sky/cli.py launch --cloud hyperstack --gpus RTX-A6000 -c test-autodown -i 1 --down echo hi
# Make sure it works with other clouds:
python sky/cli.py launch -c test-all echo hi
```

I tried to run the following, but these are failing:
```
pytest tests/test_smoke.py --hyperstack
```

cblmemo commented:
> @cblmemo - thanks for the review. There are 2 issues left:
> 
> * more descriptive error message (remove or not)
> * simplify filtering instances method
> 
> I made a few other changes:
> 
> * added type annotations where I noticed that these are missing
> * added an additional check and warning for a missing price for a VM flavor (Hyperstack added the `H200-141G-SXM5` GPU but didn't add it to the price list - https://github.com/vvidovic/skypilot/blob/a037fe9637d2f555e8c8cf01096eeaf1ee27e225/sky/clouds/service_catalog/data_fetchers/fetch_hyperstack.py#L109).
> 
> I implemented changes using a force push to my fork. This has a downside in that there is no history of changes I performed based on your review, but it makes the future merge simpler. If you don't like this approach, I will continue using normal commits and maybe rebase them at the end (for a simpler merge).

Thanks for your consideration! Yeah, force push will make review harder due to the missing history..

Actually, we have the squash & merge feature enabled in our repo, which means that you can just use normal merge and all changed will be combined into one commit to our repo. So I dont think you need to worry about that :)

cblmemo commented:
> > Thanks for adding this @vvidovic ! This is exciting. Left some discussions.
> > Also, could you help run smoke test on hyperstack to make sure it works properly?
> 
> I have run the following tests:
> 
> ```
> # Check credentials:
> python sky/cli.py check
> 
> python sky/cli.py launch -c test-single-instance --cloud hyperstack --gpus RTX-A6000 echo hi
> python sky/cli.py stop test-single-instance
> python sky/cli.py start test-single-instance
> 
> python sky/cli.py launch -c test-single-instance echo hi
> python sky/cli.py down test-single-instance
> 
> # Terminate a stopped instance:
> python sky/cli.py launch -c test-single-instance --cloud hyperstack --gpus RTX-A6000 echo hi
> python sky/cli.py stop test-single-instance; sky down test-single-instance
> 
> python sky/cli.py launch --cloud hyperstack --gpus RTX-A6000 -c test-autostop -i 1 echo hi
> python sky/cli.py launch --cloud hyperstack --gpus RTX-A6000 -c test-autodown -i 1 --down echo hi
> # Make sure it works with other clouds:
> python sky/cli.py launch -c test-all echo hi
> ```
> 
> I tried to run the following, but these are failing:
> 
> ```
> pytest tests/test_smoke.py --hyperstack
> ```
> 
> There are a lot of errors related to the storage, here is the end of the output:
> 
> ```
> ...
> ==================================================================================================== short test summary info ====================================================================================================
> FAILED tests/test_smoke.py::test_example_app - FileNotFoundError: [Errno 2] No such file or directory: 'sky'
> ...
> ERROR tests/test_smoke.py::TestStorageWithCredentials::test_public_bucket[s3://digitalcorpora-StoreType.S3] - sky.exceptions.ResourcesUnavailableError: Storage 'store: s3' specified, but AWS access is disabled. To fix, enable AWS by running `sky check`. More info: https://docs.skypilot.co/en/latest/getting-started/installation.h...
> 101 failed, 4 passed, 209 skipped, 5104 warnings, 4 errors in 7.92s
> ```
> 
> I didn't implement any storage-related functions for the Hyperstack, but I am not sure this is causing this issue.

There are some tests that requires a cloud with storage enabled, like aws or gcp. You can add a mark to skip them. Reference:

https://github.com/skypilot-org/skypilot/blob/d40331d88e023e9a888865705ad91f8b53ea35f3/tests/smoke_tests/test_sky_serve.py#L320-L321

You might also need to modify some places in the smoke test to enable hyperstack in the smoke test. You can reference to one recent PR: https://github.com/skypilot-org/skypilot/pull/4573

vvidovic-croz committed:
Replacing `assert` with `ValueError`

vvidovic-croz committed:
Description of precedence of `api_key` resolution

vvidovic-croz committed:
Hyperstack status group

Grouping statuses in 3 groups, relevant for the business logic.

vvidovic-croz committed:
Hyperstack, mounting storage not yet implemented

vvidovic-croz committed:
Hyperstack dependencies added

vvidovic-croz committed:
Hyperstack testing configuration

vvidovic-croz committed:
Fixed formatting

vvidovic-croz committed:
Fixed wrong `hyperstack` catalog name

The first letter must be lowercase.

vvidovic-croz committed:
Fixed `test_cli.py` formatting

vvidovic commented:
@cblmemo - it seems that I had issues with test configurations. I needed to add the new Hyperstack cloud in few places in test files. There was one error in catalog (`Hyperstack` vs `hyperstack`) so I hope everything works now.
Unfortunately, the Hyperstack is currently low on resources - no `RTX-A4000` or `RTX-A600` instances are currently available.
![image](https://github.com/user-attachments/assets/7d50985e-e038-47c4-bc33-16b69b303d52)

I removed those from my local catalog and the single test run was successful:
```
pytest tests/smoke_tests/test_basic.py --hyperstack -k test_minimal

D 04-22 14:45:18 skypilot_config.py:263] using default user config file: ~/.sky/config.yaml
D 04-22 14:45:18 skypilot_config.py:282] using default project config file: .sky.yaml
D 04-22 14:45:18 skypilot_config.py:303] final config: {}
bringing up nodes...
[minimal] Test started. Log: less -r /tmp/minimal-_rjiu_yy.log
[minimal] Passed.
[minimal] Log: less -r /tmp/minimal-_rjiu_yy.log
[minimal] 
.
1 passed, 656 warnings in 330.78s (0:05:30)
```

I will try to run as many tests as possible and let you know the results.

vvidovic-croz committed:
Skipping tests not available on Hyperstack

Tests skipped for Hyperstack:
- tests requiring aws or gcp
- tests using other clouds
- K80s not available
- T4 not available
- Spot instances not supported

vvidovic-croz committed:
Adding `ERROR` Hyperstack status

cblmemo commented:
> @cblmemo - it seems that I had issues with test configurations. I needed to add the new Hyperstack cloud in few places in test files. There was one error in catalog (`Hyperstack` vs `hyperstack`) so I hope everything works now. Unfortunately, the Hyperstack is currently low on resources - no `RTX-A4000` or `RTX-A600` instances are currently available. ![image](https://private-user-images.githubusercontent.com/3177210/436081108-7d50985e-e038-47c4-bc33-16b69b303d52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3OTAxOTMsIm5iZiI6MTc0NTc4OTg5MywicGF0aCI6Ii8zMTc3MjEwLzQzNjA4MTEwOC03ZDUwOTg1ZS1lMDM4LTQ3YzQtYmMzMy0xNmI2OWIzMDNkNTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjdUMjEzODEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmFiZjA5YzI0NTdkZjcyMWNjODEzZmIyZTRhMjViYmRlNzQzM2RjZWI0YmJmY2EyMWYxZTM5MzcwYTYyNGY4ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.LVw-9I4H05BGf5N4E_tc6o7HozsaAVVXeqr_dyzL_3U)
> 
> I removed those from my local catalog and the single test run was successful:
> 
> ```
> pytest tests/smoke_tests/test_basic.py --hyperstack -k test_minimal
> 
> D 04-22 14:45:18 skypilot_config.py:263] using default user config file: ~/.sky/config.yaml
> D 04-22 14:45:18 skypilot_config.py:282] using default project config file: .sky.yaml
> D 04-22 14:45:18 skypilot_config.py:303] final config: {}
> bringing up nodes...
> [minimal] Test started. Log: less -r /tmp/minimal-_rjiu_yy.log
> [minimal] Passed.
> [minimal] Log: less -r /tmp/minimal-_rjiu_yy.log
> [minimal] 
> .
> 1 passed, 656 warnings in 330.78s (0:05:30)
> ```
> 
> I will try to run as many tests as possible and let you know the results.

Thanks! Ideally all the smoke test should pass. Please ping me when all of them are resolved and I'll take another pass ;)

vvidovic-croz committed:
More robust termination logic

If we receive an error that we can't delete an instance before it
is created, retry max 18 times until it is created.

vvidovic-croz committed:
Explicit inst_id and exception mappimg

vvidovic commented:
@cblmemo - Hi, I didn't manage to go through all tests, only the `tests_basic.py`.

Only one of tests in the `tests_basic.py` is failing, I am not sure why:
```bash
pytest tests/smoke_tests/test_basic.py::test_env_check --hyperstack
```

The test fails because the Ray server doesn't start, I can't see any more info from logs:
```bash
...
D 04-28 21:12:26 provisioner.py:595] Starting Ray on the entire cluster.
Timeout after 900 seconds.
...
```

Unfortunately, when running tests, I spent more credits on VMs than I planned, and I can't spend any more time on this, at least in the near future, so I will stop working on this PR. If you think it needs more work before merging, I hope someone will take over this task.
I tested scenarios that were interesting to my business case by launching a single VM and running tasks on that VM in the Hypercloud, and that worked as expected. I don't know if something else needs more work.

Illumaria commented:
@cblmemo Hello! Do you have plans to add support for Hyperstack by merging this PR? If neither you nor @vvidovic have time to finish it, I could try and do it myself.
